tests/test-mongoc-uri.c

```diff
@@ -411,6 +418,101 @@ test_mongoc_uri_unescape (void)
 
 typedef struct
 {
+   const char         *uri;
+   bool                parses;
+   mongoc_read_mode_t  mode;
+   bson_t             *tags;
+} read_prefs_test;
+
+
+static void
+test_mongoc_uri_read_prefs (void)
+{
+   const mongoc_read_prefs_t *rp;
+   mongoc_uri_t *uri;
+   const read_prefs_test *t;
+   int i;
+
+   bson_t *tags_dcny;
+   bson_t *tags_dcny_empty;
+   bson_t *tags_dcnyusessd_dcsf_empty;
+   bson_t *tags_empty;
+
+   tags_dcny = BCON_NEW(
+      "0", "{", "dc", "ny", "}"
+   );
+
+   tags_dcny_empty = BCON_NEW(
+      "0", "{", "dc", "ny", "}",
+      "1", "{", "}"
+   );
+
+   tags_dcnyusessd_dcsf_empty = BCON_NEW(
+      "0", "{", "dc", "ny", "use", "ssd", "}",
+      "1", "{", "dc", "sf", "}",
+      "2", "{", "}"
+   );
+
+   tags_empty = BCON_NEW(
+      "0", "{", "}"
+   );
+
+   const read_prefs_test tests [] = {
+      { "mongodb://localhost/", true, MONGOC_READ_PRIMARY, NULL },
+      { "mongodb://localhost/?slaveOk=false", true, MONGOC_READ_PRIMARY, NULL },
+      { "mongodb://localhost/?slaveOk=true", true, MONGOC_READ_SECONDARY_PREFERRED, NULL },
+      { "mongodb://localhost/?readPreference=primary", true, MONGOC_READ_PRIMARY, NULL },
+      { "mongodb://localhost/?readPreference=primaryPreferred", true, MONGOC_READ_PRIMARY_PREFERRED, NULL },
+      { "mongodb://localhost/?readPreference=secondary", true, MONGOC_READ_SECONDARY, NULL },
+      { "mongodb://localhost/?readPreference=secondaryPreferred", true, MONGOC_READ_SECONDARY_PREFERRED, NULL },
+      { "mongodb://localhost/?readPreference=nearest", true, MONGOC_READ_NEAREST, NULL },
+      /* readPreference should take priority over slaveOk */
+      { "mongodb://localhost/?slaveOk=false&readPreference=secondary", true, MONGOC_READ_SECONDARY, NULL },
+      /* readPreferenceTags conflict with primary mode */
+      { "mongodb://localhost/?readPreferenceTags=", false },
+      { "mongodb://localhost/?readPreference=primary&readPreferenceTags=", false },
+      { "mongodb://localhost/?slaveOk=false&readPreferenceTags=", false },
+      { "mongodb://localhost/?readPreference=secondaryPreferred&readPreferenceTags=", true, MONGOC_READ_SECONDARY_PREFERRED, tags_empty },
+      { "mongodb://localhost/?readPreference=secondaryPreferred&readPreferenceTags=dc:ny", true, MONGOC_READ_SECONDARY_PREFERRED, tags_dcny },
+      { "mongodb://localhost/?readPreference=nearest&readPreferenceTags=dc:ny&readPreferenceTags=", true, MONGOC_READ_NEAREST, tags_dcny_empty },
+      { "mongodb://localhost/?readPreference=nearest&readPreferenceTags=dc:ny,use:ssd&readPreferenceTags=dc:sf&readPreferenceTags=", true, MONGOC_READ_NEAREST, tags_dcnyusessd_dcsf_empty },
```

Comment:
Opened [CDRIVER-598](https://jira.mongodb.org/browse/CDRIVER-598) to track this outside of this ticket/PR.